### PR TITLE
Feature/1550 primary donor

### LINKED ISSUE #1550: As a user, I'd like to have an automatically populated 'Primary Donor' lookup field on Opportunity to Contact
This field should be the contact who is checked as the Primary Opportunity Contact Role. If the Contact Role is change to another contact, a job should check for that and update the lookup field.

### LINKED ISSUE #1582: False positive from Health Check
From https://powerofus.force.com/_ui/core/chatter/groups/GroupProfilePage?g=0F980000000CjRe&fId=0D580000028OzMD

User is getting error message with contents of label healthSolutionScheduleJobs.

The functionality that notifies of this issue has not changed since almost a year ago (see line 503 in https://github.com/SalesforceFoundation/Cumulus/blame/104dbfac7bcc4aad4b35030edd98039dcdf9fe36/src/classes/STG_PanelHealthCheck_CTRL.cls). However, recent commit added one more class to be scheduled (https://github.com/SalesforceFoundation/Cumulus/commit/ce18095683b88894df7ac6c1e0d6342c08f6276e), which is causing to miscount. @njjc

### PR DESCRIPTION
**Pull request by njjc**
# Warning
# Info
- This release includes an oft requested feature: a Primary Contact lookup from Opportunity! This field automatically creates and updates the primary Opportunity Contact Role.

# Issues
Fixes #1550
Fixes #1582

### CODE REVIEW

File: src/classes/OPP_OpportunityContactRoles_TDTM.cls [hunk 1]
```diff
@@ -36,144 +36,199 @@
...
+        	createOrUpdateContactRoles((list<Opportunity>) newlist);
+        } else if (triggerAction == TDTM_Runnable.Action.AfterUpdate) {
+            updateContactRoles((list<Opportunity>) newlist, (list<Opportunity>) oldlist);
         }
-        
-        return dmlWrapper;
+
+        // we can't defer the dml.  too many of our other opp related triggers depend on the ocr's being saved.
+        // disable TDTM so we don't waste resources running through opp triggers a second time while fixing Primary Contact lookup
+        if (!dmlWrapper.objectsToInsert.isEmpty() || !dmlWrapper.objectsToUpdate.isEmpty() || !dmlWrapper.objectsToDelete.isEmpty() || !dmlWrapper.objectsToUndelete.isEmpty()) {
```
Comment (davidhabib) [inline src/classes/OPP_OpportunityContactRoles_TDTM.cls]: I think you could remove the if testing, and just let processDML() do this.

File: src/classes/OPP_OpportunityContactRoles_TDTM.cls [hunk 2]
```diff
@@ -36,144 +36,199 @@
...
 
         List<OpportunityContactRole> CRs = new List<OpportunityContactRole>();
 
         //Straightforward creation of opp contact role.      
-        for(Opportunity o : opportunities){
-            if(o.npe01__Contact_ID_for_Role__c != null) {
-                try {
-                    CRs.add(new OpportunityContactRole(OpportunityId = o.Id, ContactId = o.npe01__Contact_ID_for_Role__c, Role = getDefaultRole(), IsPrimary = true));
-                } catch (exception e) {
-                    o.addError(Label.npe01.Opportunity_Contact_Role_Error_Bad_Contact_Id);
```
Comment (davidhabib) [inline src/classes/OPP_OpportunityContactRoles_TDTM.cls]: So this is where we would have generated an error about invalid contact id's.  are we now preventing this from happening?  ie, progagating an error to the integration (or more likely person importing data) that they set a bogus id.

Comment (njjc) [inline src/classes/OPP_OpportunityContactRoles_TDTM.cls]: We are preventing this now. I'll put the addError back into the catch so it propagates.

**Comment by njjc** [comment]
@davidhabib ready for review, and merge once build passes. Thanks!

**Pull request merged**